src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java

```diff
@@ -356,16 +384,20 @@ private void setCachedResultAndResumeToken() {
         return context != null ? context.offsetStorageReader().offset(createPartitionMap(sourceConfig)) : null;
     }
 
-    private BsonDocument getResumeAfter(final MongoSourceConfig cfg) {
-        return getResumeAfter(getOffset(cfg));
+    private BsonDocument getResumeToken(final MongoSourceConfig cfg) {
+        return getResumeToken(getOffset(cfg));
     }
 
-    private BsonDocument getResumeAfter(final Map<String, Object> offset) {
-        if (cachedResumeAfter != null) {
-            return cachedResumeAfter;
+    private BsonDocument getResumeToken(final Map<String, Object> offset) {
+        BsonDocument resumeToken = null;
+        if (cachedResumeToken != null) {
+            resumeToken = cachedResumeToken;
+            cachedResumeToken = null;
+        } else if (invalidatedCursor) {
+            invalidatedCursor = false;
```

Comment:
If the cursor was invalidated don't try and use the saved offset as that will be from before the invalidation event (because any seen invalidation events would have cached the resumeToken).